gluonnlp/data/stream.py

```diff
@@ -0,0 +1,313 @@
+            for idx in iter(sampler(num_samples)):
+                yield corpus[idx]
+
+class LanguageModelStream(CorpusStream):
+    """Streams a corpus consisting of multiple text files that match provided
+    `file_pattern`, and produces a language modeling stream given the provided
+    sample splitter and word tokenizer.
+
+    Parameters
+    ----------
+    file_pattern: str
+        Path to the input text files.
+    encoding : str, default 'utf8'
+        File encoding format.
+    skip_empty : bool, default True
+        Whether to skip the empty samples produced from sample_splitters. If False, `bos` and `eos`
+        will be added in empty samples.
+    sample_splitter : function, default str.splitlines
+        A function that splits the dataset string into samples.
+    tokenizer : function or None, default str.split
+        A function that splits each sample string into list of tokens. If None, raw samples are
+        returned according to `sample_splitter`.
+    bos : str or None, default None
+        The token to add at the begining of each sequence. If None, or if tokenizer is not
+        specified, then nothing is added.
+    eos : str or None, default None
+        The token to add at the end of each sequence. If None, or if tokenizer is not
+        specified, then nothing is added.
+    sampler : str, {'sequential', 'random'}, defaults to 'random'
+        The sampler used to sample texts within a file.
+
+        - 'sequential': SequentialSampler
+        - 'random': RandomSampler
+    file_sampler : str, {'sequential', 'random'}, defaults to 'random'
+        The sampler used to sample a file.
+
+        - 'sequential': SequentialSampler
+        - 'random': RandomSampler
+    """
+    def __init__(self, file_pattern, encoding='utf8', skip_empty=True,
+                 sample_splitter=lambda s: s.splitlines(), tokenizer=lambda s: s.split(),
+                 bos=None, eos=None, sampler='random', file_sampler='random'):
+        self._file_pattern = file_pattern
+        self._encoding = encoding
+        self._skip_empty = skip_empty
+        self._sample_splitter = sample_splitter
+        self._tokenizer = tokenizer
+        self._bos = bos
+        self._eos = eos
+        self._sampler = sampler
+        self._file_sampler = file_sampler
+        super(LanguageModelStream, self).__init__(self._file_pattern, flatten=True,
+                                                  encoding=self._encoding,
+                                                  skip_empty=self._skip_empty,
+                                                  sample_splitter=self._sample_splitter,
+                                                  tokenizer=self._tokenizer, bos=self._bos,
+                                                  eos=self._eos, sampler=self._sampler,
+                                                  file_sampler=self._file_sampler)
+
+    def bptt_batchify(self, vocab, seq_len, batch_size, last_batch='keep'):
```

Comment:
should we provide batchify-only here? Then we allow users to play with the bptt in user-defined get_batch. I know it is not essential for now since the training and inference should use random sampling. But could be a useful feature.

Reply:
Discussed offline, not applicable to streams